src/main/java/com/avispl/dal/communicator/pexip/PexipManagementNode.java

```diff
@@ -0,0 +1,929 @@
+     */
+    private void sendReportsEmail(List<ReportWrapper> reports) throws MessagingException, IOException {
+        List<File> files = new ArrayList<>();
+        try {
+            MimeMessageHelper helper = prepareMimeMessageHelper("Reports");
+
+            for (ReportWrapper reportWrapper : reports) {
+                String reportFileName = reportWrapper.getReportName() + ".csv";
+                File file = new File(reportFileName);
+                files.add(file);
+                BufferedWriter writer = new BufferedWriter(new FileWriter(reportFileName));
+
+                List<Map<String, String>> reportsList = reportWrapper.getReport();
+
+                Optional<Map<String, String>> map = reportsList.stream().max(Comparator.comparing(Map::size));
+                String keys = "";
+                if (map.isPresent()) {
+                    keys = String.join(",", map.get().keySet());
+                    writer.append(keys);
+                }
+
+                for (Map<String, String> stringStringMap : reportsList) {
+                    writer.newLine();
+
+                    Arrays.stream(keys.split(",")).forEach(s -> {
+                        if (!stringStringMap.containsKey(s)) {
+                            stringStringMap.put(s, "-");
+                        }
+                    });
+                    writer.append(String.join(",", stringStringMap.values()));
+                }
+                writer.close();
+                helper.addAttachment(file.getName(), file);
+            }
+
+            mailSender.send(helper.getMimeMessage());
+        } finally {
+            files.forEach(File::deleteOnExit);
+        }
+    }
+
+    /**
+     * Uptime is received in seconds, need to normalize it and make it human readable, like
+     * 1 day(s) 5 hour(s) 12 minute(s) 55 minute(s)
+     * Incoming parameter is may have a decimal point, so in order to safely process this - it's rounded first.
+     * We don't need to add a segment of time if it's 0.
+     *
+     * @param uptimeSeconds value in seconds
+     * @return string value of format 'x day(s) x hour(s) x minute(s) x minute(s)'
+     */
+    private String normalizeUptime(long uptimeSeconds) {
+        StringBuilder normalizedUptime = new StringBuilder();
+
+        long seconds = uptimeSeconds % 60;
+        long minutes = uptimeSeconds % 3600 / 60;
+        long hours = uptimeSeconds % 86400 / 3600;
+        long days = uptimeSeconds / 86400;
+
+        if (days > 0) {
+            normalizedUptime.append(days).append(" day(s) ");
+        }
+        if (hours > 0) {
+            normalizedUptime.append(hours).append(" hour(s) ");
+        }
+        if (minutes > 0) {
+            normalizedUptime.append(minutes).append(" minute(s) ");
+        }
+        if (seconds > 0) {
+            normalizedUptime.append(seconds).append(" second(s)");
+        }
+        return normalizedUptime.toString().trim();

```

Comment:
I think I've seen this method get used in multiple adapters...maybe worth putting in a utility class (just a suggestion)

Reply:
Yes, there's a bunch of methods like this and it's planned to make them a part of DAL core utils :) (along with controls instantiation)